### PR TITLE
Users can now create custom webhooks.

### PR DESCRIPTION
**Pull request by spitfire305**

### CODE REVIEW

File: app/Http/Requests/StoreWebhook.php [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    /**
+     * Get the validation rules that apply to the request.
+     *
+     * @return array
+     */
+    public function rules()
+    {
+        return [
+            'action'    => 'integer|required',
+            'url'       => 'string|required',
```
Comment (ilestis) [inline app/Http/Requests/StoreWebhook.php:28]: Please also add https://laravel.com/docs/10.x/validation#rule-active-url, and a max length to avoid mysql errors

File: app/Http/Resources/EntityResource.php [hunk 2]
```diff
@@ -48,8 +48,10 @@ public function toArray($request)
     {
         /** @var \App\Models\Entity $entity */
         $entity = $this->resource;
-
-        $url = $entity->url();
+        $url = '';
+        if (CampaignLocalization::getCampaign()) {
```
Comment (ilestis) [inline app/Http/Resources/EntityResource.php:52]: Why is this change needed? Does this mean that there is no url in the payload? Because that's probably one of the most useful fields along with the name

File: app/Jobs/EntityWebhookJob.php [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+{
+    use Dispatchable;
+    use InteractsWithQueue;
+    use Queueable;
+    use SerializesModels;
+
+    /**
+     * @var int
+     */
+    public $campaignId;
```
Comment (ilestis) [inline app/Jobs/EntityWebhookJob.php:30]: Variables can be typed so that we don't need to bother with the phpdoc comments

File: app/Jobs/EntityWebhookJob.php [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+    /**
+     * Execute the job.
+     *
+     * @return void
+     */
+    public function handle()
+    {
+        Log::info('EntityWebhookJob for entity #' . $this->entity->id);
+
+        $webhooks = Webhook::where('campaign_id', $this->campaignId)->where('action', $this->action)->where('status', 1)->get();
```
Comment (ilestis) [inline app/Jobs/EntityWebhookJob.php:78]: Probably worth doing a scope with two params (campaign id and action)

Comment (ilestis) [inline app/Jobs/EntityWebhookJob.php:78]: Also missing with() statements, please enable lazy loading alerts :-)

File: app/Jobs/EntityWebhookJob.php [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+        foreach ($webhooks as $webhook) {
+            if ($webhook->tags()->count() > 0) {
+                $tags = $webhook->tags()->pluck('tags.id')->all();
+
+                if (!$this->entity->tags()->whereIn('tags.id', $tags)->first()) {
```
Comment (ilestis) [inline app/Jobs/EntityWebhookJob.php:84]: Is this setup so that entities need _all the tags_ or _at least one of the tags_? Because whichever version it is, we need to warn users about it or they'll get confused and spam the discord with the same question.

File: app/Jobs/EntityWebhookJob.php [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+     * @return void
+     */
+    public function handle()
+    {
+        Log::info('EntityWebhookJob for entity #' . $this->entity->id);
+
+        $webhooks = Webhook::where('campaign_id', $this->campaignId)->where('action', $this->action)->where('status', 1)->get();
+
+        foreach ($webhooks as $webhook) {
+            if ($webhook->tags()->count() > 0) {
```
Comment (ilestis) [inline app/Jobs/EntityWebhookJob.php:81]: This whole if could be a function on the job 
```php
if ($this->invalidTags($webhook)) {
   continue;
}

//
protected function invalidTags(Webhook $webhook): bool
{

}
```

File: app/Models/Relations/CampaignRelations.php [hunk 7]
```diff
@@ -114,6 +115,11 @@ public function roles(): HasMany
         return $this->hasMany(CampaignRole::class);
     }
 
+    public function webhooks(): HasMany
```
Comment (ilestis) [inline app/Models/Relations/CampaignRelations.php:118]: Missing the `@property Collection|Webhook[] $webhooks` auto-complete

File: app/Observers/EntityObserver.php [hunk 8]
```diff
@@ -187,6 +188,7 @@ public function created(Entity $entity)
         if (!request()->has('attr_name')) {
             $this->attributeService->applyEntityTemplates($entity);
         }
+        EntityWebhookJob::dispatch($entity, $entity->campaign, auth()->user(), 1);
```
Comment (ilestis) [inline app/Observers/EntityObserver.php:191]: Type could be a new Enum (app/Enums) to make it a bit easier to read

File: app/Observers/EntityObserver.php [hunk 9]
```diff
@@ -197,6 +199,7 @@ public function updated(Entity $entity)
 
         // Queue job when an entity was updated
         EntityUpdatedJob::dispatch($entity);
+        EntityWebhookJob::dispatch($entity, $entity->campaign, auth()->user(), 2);
```
Comment (ilestis) [inline app/Observers/EntityObserver.php:202]: Since campaign is taken from the entity, the whole param can be removed from the code and job

File: app/Renderers/Layouts/Campaign/Webhook.php [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+    /**
+     * Available columns
+     * @return array[]
+     */
+    public function columns(): array
+    {
+        $columns = [
+            'action' => [
+                'key' => 'action',
+                'label' => 'campaigns.webhooks.fields.event',
```
Comment (ilestis) [inline app/Renderers/Layouts/Campaign/Webhook.php:18]: For our own sanity, please move all the labels to campaigns/webooks.php instead of inside of the already too big campaigns.php translation file

File: database/migrations/2024_03_27_173140_create_webhooks.php [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+            $table->unsignedTinyInteger('action');
+            $table->string('url', 191);
+            $table->unsignedTinyInteger('type');
+            $table->text('message')->nullable();
+            $table->timestamps();
+            $table->unsignedInteger('created_by')->nullable();
+            $table->unsignedInteger('updated_by')->nullable();
+            $table->unsignedTinyInteger('status');
+
+            $table->foreign('created_by')->references('id')->on('users')->onDelete('set null');
```
Comment (ilestis) [inline database/migrations/2024_03_27_173140_create_webhooks.php:25]: onDelete('set null') can be replace by nullOnDelete() which is easier on the eyes

File: lang/en/campaigns.php [hunk 12]
```diff
@@ -480,4 +481,59 @@
         'public'    => 'Public campaign',
         'review'    => 'Awaiting Review',
     ],
+    'webhooks'              => [
```
Comment (ilestis) [inline lang/en/campaigns.php:484]: Please move it all to a new file `campaigns/webhooks.php` to keep translation files smaller

File: resources/views/campaigns/webhooks/_form.blade.php [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    >
+    {!! Form::select('action', [1 => __('campaigns.webhooks.fields.events.new'), 2 => __('campaigns.webhooks.fields.events.edited'), 3 => __('campaigns.webhooks.fields.events.deleted')]) !!}
+</x-forms.field>
+
+<x-forms.field
+    field="url"
+    :required="true"
+    :label="__('campaigns.webhooks.fields.url')"
+    >
+    {!! Form::text('url', null, ['placeholder' => __('campaigns.webhooks.placeholders.url'), 'class' => '', 'maxlength' => 200, 'required']) !!}
```
Comment (ilestis) [inline resources/views/campaigns/webhooks/_form.blade.php:16]: max-length is different than the value in the database and will cause errors/data loss

File: resources/views/campaigns/webhooks/index.blade.php [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+        <a href="{{ route('webhooks.create', $campaign) }}" class="btn2 btn-primary btn-sm"
+           data-url="{{ route('webhooks.create', $campaign) }}"
+        >
+            <x-icon class="plus"></x-icon>
+            {{ __('campaigns.webhooks.actions.add') }}
+        </a>
+    @endif
+    @endif
+</div>
+@if (!$campaign->boosted())
```
Comment (ilestis) [inline resources/views/campaigns/webhooks/index.blade.php:21]: Should test for premium campaigns, not boosted

File: resources/views/campaigns/webhooks/index.blade.php [hunk 15]
```diff
@@ -0,0 +1,48 @@
+<div class="flex gap-2 items-center">
+    <h3 class="inline-block grow">
+        {{ __('campaigns.show.tabs.webhooks') }}</span>
+    </h3>
+    @if ($campaign->boosted())
```
Comment (ilestis) [inline resources/views/campaigns/webhooks/index.blade.php:5]: Should test for premium campaigns, not the old boosted ones

File: routes/campaigns/campaign.php [hunk 16]
```diff
@@ -173,3 +173,15 @@
 
 Route::get('/w/{campaign}/confirm-delete', [App\Http\Controllers\ConfirmController::class, 'index'])->name('confirm-delete');
 Route::post('/w/{campaign}/vanity-validate', [App\Http\Controllers\Campaign\VanityController::class, 'index'])->name('campaign.vanity-validate');
+
+// Permission save
+Route::get('/w/{campaign}/webhooks', [App\Http\Controllers\Campaign\WebhookController::class, 'index'])->name('webhooks.index');
+Route::get('/w/{campaign}/webhooks/create', [App\Http\Controllers\Campaign\WebhookController::class, 'create'])->name('webhooks.create');
```
Comment (ilestis) [inline routes/campaigns/campaign.php:179]: Delete most of these by using the `Route::resources([` code at the beginning of this file

File: app/Jobs/EntityWebhookJob.php [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+    }
+
+    /**
+     * Execute the job.
+     *
+     * @return void
+     */
+    public function handle()
+    {
+        Log::info('EntityWebhookJob for entity #' . $this->entity->id);
```
Comment (ilestis) [inline app/Jobs/EntityWebhookJob.php:76]: Missing check to see if the campaign is still premium

**Review by ilestis — CHANGES_REQUESTED** [review]
Good job for a first draft!

File: app/Jobs/EntityWebhookJob.php (hunk 5 above)
Comment (spitfire305) [reply app/Jobs/EntityWebhookJob.php:84]: At least one of the tags.

File: app/Enums/WebhookAction.php [hunk 18]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+namespace App\Enums;
+
+enum WebhookAction: int
+{
+    case NEW = 1;
```
Comment (ilestis) [inline app/Enums/WebhookAction.php:7]: Please use CREATED so it follows the same convention as the other two options :P

File: app/Http/Controllers/Campaign/WebhookController.php [hunk 19]
```diff
@@ -64,7 +64,7 @@ public function create(Campaign $campaign)
     {
         $this->authorize('webhooks', $campaign);
 
-        return view('campaigns.webhooks.create', ['campaign' => $campaign, 'model' => $campaign]);
+        return view('campaigns/webhooks.create', ['campaign' => $campaign, 'model' => $campaign]);
```
Comment (ilestis) [inline app/Http/Controllers/Campaign/WebhookController.php:67]: view paths always use . Also, why is model set to campaign? That seems very wrong, it should be undefined

File: app/Http/Controllers/Campaign/WebhookController.php [hunk 20]
```diff
@@ -79,14 +79,14 @@ public function store(StoreWebhook $request, Campaign $campaign)
...
-            ->with('success_raw', __('campaigns.webhooks.create.success'));
+            ->with('success_raw', __('campaigns/webhooks.create.success'));
     }
 
     public function edit(Campaign $campaign, Webhook $webhook)
     {
         $this->authorize('webhooks', $campaign);
 
-        return view('campaigns.webhooks.edit', [
+        return view('campaigns/webhooks.edit', [
```
Comment (ilestis) [inline app/Http/Controllers/Campaign/WebhookController.php:89]: . for view paths

File: app/Http/Controllers/Campaign/WebhookController.php [hunk 21]
```diff
@@ -79,14 +79,14 @@ public function store(StoreWebhook $request, Campaign $campaign)
         Webhook::create($data);
 
         return redirect()->route('webhooks.index', $campaign)
-            ->with('success_raw', __('campaigns.webhooks.create.success'));
+            ->with('success_raw', __('campaigns/webhooks.create.success'));
```
Comment (ilestis) [inline app/Http/Controllers/Campaign/WebhookController.php:82]: You only need to use success_raw when you have a link in the success message. Otherwise just `success` is fine

File: app/Http/Controllers/Campaign/WebhookController.php [hunk 22]
```diff
@@ -99,7 +99,7 @@ public function update(StoreWebhook $request, Campaign $campaign, Webhook $webho
 
         $webhook->update($request->all());
         return redirect()->route('webhooks.index', $campaign)
-            ->with('success_raw', __('campaigns.webhooks.edit.success'));
+            ->with('success_raw', __('campaigns/webhooks.edit.success'));
```
Comment (ilestis) [inline app/Http/Controllers/Campaign/WebhookController.php:102]: Same here

File: app/Http/Controllers/Campaign/WebhookController.php [hunk 23]
```diff
@@ -109,14 +109,14 @@ public function destroy(Campaign $campaign, Webhook $webhook)
         $webhook->delete();
 
         return redirect()->route('webhooks.index', $campaign)
-            ->with('success_raw', __('campaigns.webhooks.destroy.success'));
+            ->with('success_raw', __('campaigns/webhooks.destroy.success'));
```
Comment (ilestis) [inline app/Http/Controllers/Campaign/WebhookController.php:112]: Same same

File: app/Http/Controllers/Campaign/WebhookController.php [hunk 24]
```diff
@@ -109,14 +109,14 @@ public function destroy(Campaign $campaign, Webhook $webhook)
...
     }
 
     public function status(Campaign $campaign, Webhook $webhook)
     {
         $this->authorize('webhooks', $campaign);
 
-        return view('campaigns.webhooks.status', [
+        return view('campaigns/webhooks.status', [
```
Comment (ilestis) [inline app/Http/Controllers/Campaign/WebhookController.php:119]: Path .

**Review by ilestis — COMMENTED** [review]
Also check the code saving tags as I've refactored that in develop and so we no longer need the copy-pasted loops

**Review by ilestis — APPROVED** [review]
Nice work!

File: app/Services/Entity/TagService.php [hunk 25]
```diff
@@ -58,7 +61,13 @@ public function create(mixed $name): Tag
         $tag = new Tag([
             'name' => Purify::clean($name),
         ]);
-        $tag->campaign_id = isset($this->campaign) ? $this->campaign->id : $this->entity->campaign_id;
+
```
Comment (ilestis) [inline app/Services/Entity/TagService.php:64]: Webhooks aren't allowed to create new tags so all of this can be reset

**Review by ilestis — CHANGES_REQUESTED** [review]
Revert changes

**Review by ilestis — APPROVED** [review]
Gg ez clap no cap
